crates/viewer/re_renderer/src/video/player.rs

```diff
@@ -129,64 +145,90 @@ impl VideoPlayer {
-                self.video_texture.frame_info = None;
-                true
-            } else if presentation_timestamp < time_range.start {
-                // We're seeking backwards and somehow forgot to reset.
-                true
-            } else if presentation_timestamp < time_range.end {
-                false // it is an active frame
-            } else if let Some(timescale) = video_description.timescale {
-                let how_outdated = presentation_timestamp - time_range.end;
-                if how_outdated.duration(timescale) < DECODING_GRACE_DELAY {
-                    false // Just outdated by a little bit - show no spinner
-                } else {
-                    true // Very old frame - show spinner
+        // Find which sample best represents the requested PTS.
+        let requested_sample_idx = video_description
+            .latest_sample_index_at_presentation_timestamp(requested_pts)
+            .ok_or(VideoPlayerError::EmptyVideo)?;
+        let requested_sample = video_description.samples.get(requested_sample_idx); // This is only `None` if we no longer have the sample around.
+        let requested_pts = requested_sample
+            .map(|s| s.presentation_timestamp)
+            .unwrap_or(requested_pts);
+
+        // Ensure we have enough samples enqueued to the decoder to cover the request.
+        // (This method also makes sure that the next few frames become available, so call this even if we already have the frame we want.)
+        self.enqueue_samples(video_description, requested_sample_idx, video_buffers)?;
+
+        // Grab best decoded frame for the requested PTS and discard all earlier frames to save memory.
+        if let Some(decoded_frame) = self
+            .sample_decoder
+            .latest_decoded_frame_at_and_drop_earlier_frames(requested_pts)
+        {
+            // Update the texture only if:
+            // * we're not already up to date.
+            let current_frame_info = self.video_texture.frame_info.as_ref();
+            let outdated_frame =
+                current_frame_info.is_none_or(|info| info.presentation_timestamp != requested_pts);
+
+            // * the decoded frame isn't too far behind.
+            //   This happens when catching up on a seek, very rapid playback or simply too slow decoding.
+            //   Without this tolerance, we'd get a "rubber-band effect" in playback where we show a lot of outdated frames before (hopefully) catching up.
+            //   This is especially common when seeking to the end of a large GOP since the decoder has to start from its beginning.
+            let timescale = video_description.timescale.unwrap_or(Timescale::new(30)); // Assume 30 time units per second if there's no scale. As good as any guess!;
+            let pts_diff = decoded_frame.info.presentation_timestamp - requested_pts;
+            let not_too_far_behind = pts_diff.duration(timescale) <= TOLLERATED_OUTPUT_DELAY;
```

Comment:
Is there a risk `not_too_far_behind` will always be `false` if e.g. the software decoder takes more than TOLLERATED_OUTPUT_DELAY to decode a big 8k frame, or whatever?

Reply:
had a long conversation about this. TODO:
* make this stateful - enter "outdated state" on tolerance exceeding, exit it when caught up (fully!)
* comment things:
    * suggest adding wallclock timers for tracking the time it takes for decoding individual frames (for debugging)
    * .. some stuff we don't do and why
* can we do this is number of frames? that seems to be more wayyy more robust!